jsaddle/src/Language/Javascript/JSaddle/Types.hs

```diff
@@ -760,9 +769,24 @@ instance MonadError JSVal JSM where
     finishVar <- JSM $ liftIO newEmptyMVar
     asyncExceptionVar <- JSM $ liftIO newEmptyMVar
     JSM $ liftIO $ atomically $ modifyTVar' tries $ M.insert tryId finishVar
-    env <- JSM ask
-    let end = sendReq Req_FinishTry
-        action = (runReaderT (unJSM (a <* end)) $ env { _jsContextRef_myTryId = tryId })
+    env <- JSM $ do
+      oldEnv <- ask
+      let newEnv = oldEnv { _jsContextRef_myTryId = tryId }
+      case _jsContextRef_syncThreadId oldEnv of
+        Nothing -> pure newEnv
+        Just _ -> do
+          v <- liftIO $ newTVarIO []
+          pure $ newEnv { _jsContextRef_waitForResults = Just v }
+    let end = do
+          JSM $ (asks _jsContextRef_waitForResults >>=) $ mapM_ $ \tVar -> do
+              vars <- liftIO $ readTVarIO tVar
+              -- wait for results of these
+              mapM_ (\jsVal -> case getPrimJSVal jsVal of
+                       PrimVal_Undefined -> pure ()
```

Comment:
Shouldn't this be different from the default `pure ()`?

Reply:
The intention here is force the evaluation of `case`. I think there are more idiomatic ways of achieving that.
But there is also a bug here, the `waitForResults` should have `Weak` pointer to the `JSVal`, and it only need to wait if it is alive at this point.

Reply:
I am now using `Weak`